File: yocton.c

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>

#include "yocton.h"

#define ERROR_BUF_SIZE 100

#define ERROR_ALLOC "memory allocation failure"
#define ERROR_EOF   "unexpected EOF"

#define CHECK_OR_GOTO_FAIL(condition) \
	if (!(condition)) { goto fail; }

enum token_type {
	TOKEN_STRING,
	TOKEN_COLON,
	TOKEN_OPEN_BRACE,
	TOKEN_CLOSE_BRACE,
	TOKEN_EOF,
	TOKEN_ERROR,
};

struct yocton_instream {
	// callback gets invoked to read more data from input.
	yocton_read callback;
	void *callback_handle;
	// buf is the input buffer containing last data read by callback.
	// buf[buf_offset:buf_len] is still to read.
	char *buf;
	size_t buf_len, buf_size;
	unsigned int buf_offset;
	// string contains the last string token read.
	char *string;
	size_t string_len, string_size;
	// error_buf is non-empty if an error occurs during parsing.
	char *error_buf;
	int lineno;
	struct yocton_object *root;
};

static void input_error(struct yocton_instream *s, char *fmt, ...)
{
	va_list args;

	// We only store the first error.
	if (strlen(s->error_buf) > 0) {
		return;
	}
	va_start(args, fmt);
	vsnprintf(s->error_buf, ERROR_BUF_SIZE, fmt, args);
}

// Assign the result of an allocation, storing an error if result == NULL.
static int assign_alloc(void *ptr, struct yocton_instream *s, void *result)
{
	if (result == NULL) {
		input_error(s, ERROR_ALLOC);
		return 0;
	}
	* ((void **) ptr) = result;
	return 1;
}

static int peek_next_char(struct yocton_instream *s, char *c)
{
	if (s->buf_offset >= s->buf_len) {
		s->buf_len = s->callback(s->buf, s->buf_size,
		                         s->callback_handle);
		if (s->buf_len == 0) {
			return 0;
		}
		s->buf_offset = 0;
	}
	*c = s->buf[s->buf_offset];
	return 1;
}

// Read next char from input. Failure stores an error.
static int read_next_char(struct yocton_instream *s, char *c)
{
	if (!peek_next_char(s, c)) {
		input_error(s, ERROR_EOF);
		return 0;
	}
	++s->buf_offset;
	if (*c == '\n') {
		++s->lineno;
	}
	return 1;
}

static int is_bare_string_char(int c)
{
       return isalnum(c) || strchr("_-+.", c) != NULL;
}

static int append_string_char(struct yocton_instream *s, char c)
{
	if (s->string_len + 1 >= s->string_size) {
		s->string_size = s->string_size == 0 ? 64 : s->string_size * 2;
		CHECK_OR_GOTO_FAIL(
		    assign_alloc(&s->string, s,
		        realloc(s->string, s->string_size)));
	}
	s->string[s->string_len] = c;
	s->string[s->string_len + 1] = '\0';
	++s->string_len;
	return 1;
fail:
	return 0;
}

static int unescape_string_char(char *c) {
	switch (*c) {
		case 'a':  *c = '\a'; return 1;
		case 'b':  *c = '\b'; return 1;
		case 'n':  *c = '\n'; return 1;
		case 'r':  *c = '\r'; return 1;
		case 't':  *c = '\b'; return 1;
		case '\\': *c = '\\'; return 1;
		case '\'': *c = '\''; return 1;
		case '"':  *c = '\"'; return 1;
		default: return 0;
		// TODO: \x, \u etc.
	}
}

// Read quote-delimited "C style" string.
static enum token_type read_string(struct yocton_instream *s)
{
	char c;
	s->string_len = 0;
	for (;;) {
		CHECK_OR_GOTO_FAIL(read_next_char(s, &c));
		if (c == '"') {
			return TOKEN_STRING;
		} else if (c == '\\') {
			CHECK_OR_GOTO_FAIL(read_next_char(s, &c));
			if (!unescape_string_char(&c)) {
				input_error(s, "unknown string escape: \\%c", c);
				return TOKEN_ERROR;
			}
		}
		CHECK_OR_GOTO_FAIL(append_string_char(s, c));
	}
fail:
	return TOKEN_ERROR;
}

static enum token_type read_bare_string(struct yocton_instream *s, char first)
{
	char c;
	if (!is_bare_string_char(first)) {
		input_error(s, "unknown token: not valid bare-string character");
		return TOKEN_ERROR;
	}
	s->string_len = 0;
	CHECK_OR_GOTO_FAIL(append_string_char(s, first));
	// Reaching EOF in the middle of the string is explicitly okay here:
	while (peek_next_char(s, &c) && is_bare_string_char(c)) {
		CHECK_OR_GOTO_FAIL(read_next_char(s, &c));
		CHECK_OR_GOTO_FAIL(append_string_char(s, c));
	}
	return TOKEN_STRING;
fail:
	return TOKEN_ERROR;
}

static enum token_type read_next_token(struct yocton_instream *s)
{
	char c;
	if (strlen(s->error_buf) > 0) {
		return TOKEN_ERROR;
	}
	// Skip past any spaces. Reaching EOF is not always an error.
	do {
		if (!peek_next_char(s, &c)) {
			return TOKEN_EOF;
		}
		CHECK_OR_GOTO_FAIL(read_next_char(s, &c));
	} while (isspace(c));

	switch (c) {
		case ':':  return TOKEN_COLON;
		case '{':  return TOKEN_OPEN_BRACE;
		case '}':  return TOKEN_CLOSE_BRACE;
		case '\"': return read_string(s);
		default:   return read_bare_string(s, c);
	}
fail:
	return TOKEN_ERROR;
}

struct yocton_object {
	struct yocton_instream *instream;
	struct yocton_field *field;
	int done;
};

struct yocton_field {
	enum yocton_field_type type;
	char *name, *value;
	struct yocton_object *parent, *child;
};

static void free_obj(struct yocton_object *obj);

static void free_field(struct yocton_field *field)
{
	if (field == NULL) {
		return;
	}
	free_obj(field->child);
	field->child = NULL;

	free(field->name);
	free(field->value);
	free(field);
}

static void free_obj(struct yocton_object *obj)
{
	if (obj == NULL) {
		return;
	}
	free_field(obj->field);
	obj->field = NULL;
	free(obj);
}

static void free_instream(struct yocton_instream *instream)
{
	if (instream == NULL) {
		return;
	}
	free(instream->buf);
	free(instream->error_buf);
	free(instream->string);
	free(instream);
}

static size_t fread_wrapper(void *buf, size_t buf_size, void *handle)
{
	return fread(buf, 1, buf_size, (FILE *) handle);
}

struct yocton_object *yocton_read_from(FILE *fstream)
{
	return yocton_read_with(fread_wrapper, fstream);
}

struct yocton_object *yocton_read_with(yocton_read callback, void *handle)
{
	struct yocton_instream *instream = NULL;
	struct yocton_object *obj = NULL;

	instream = calloc(1, sizeof(struct yocton_instream));
	CHECK_OR_GOTO_FAIL(instream != NULL);
	obj = calloc(1, sizeof(struct yocton_object));
	CHECK_OR_GOTO_FAIL(obj != NULL);

	obj->instream = instream;
	obj->field = NULL;
	obj->done = 0;

	instream->root = obj;
	instream->callback = callback;
	instream->callback_handle = handle;
	instream->lineno = 1;
	instream->buf_len = 0;
	instream->buf_offset = 0;
	instream->buf_size = 256;
	CHECK_OR_GOTO_FAIL(
	    assign_alloc(&instream->error_buf, instream,
	        calloc(ERROR_BUF_SIZE, 1)));
	CHECK_OR_GOTO_FAIL(
	    assign_alloc(&instream->buf, instream,
	        calloc(instream->buf_size, sizeof(char))));
	instream->string_size = 0;
	instream->string = NULL;

	return obj;

fail:
	free_instream(instream);
	free_obj(obj);
	return NULL;
}

void yocton_free(struct yocton_object *obj)
{
	if (obj != obj->instream->root) {
		return;
	}

	free_instream(obj->instream);
	free_obj(obj);
}

// If we're partway through reading a child object, skip through any
// of its fields so we can read the next of ours.
static void skip_forward(struct yocton_object *obj)
{
	struct yocton_object *child;
	if (obj->field == NULL || obj->field->child == NULL) {
		return;
	}
	child = obj->field->child;
	// Read out all subfields until we get a NULL response and have
	// finished skipping over them.
	while (yocton_next_field(child) != NULL);
	free_obj(child);
	obj->field->child = NULL;
}

static struct yocton_field *next_field(struct yocton_object *obj)
{
	struct yocton_field *f = NULL;

	CHECK_OR_GOTO_FAIL(
	    assign_alloc(&f, obj->instream,
	        calloc(1, sizeof(struct yocton_field))));
	obj->field = f;
	f->parent = obj;
	CHECK_OR_GOTO_FAIL(
	    assign_alloc(&f->name, obj->instream,
	        strdup(obj->instream->string)));

	switch (read_next_token(obj->instream)) {
		case TOKEN_COLON:
			// This is the string:string case.
			f->type = YOCTON_FIELD_STRING;
			if (read_next_token(obj->instream) != TOKEN_STRING) {
				input_error(obj->instream, "string expected "
				            "to follow ':'");
				goto fail;
			}
			CHECK_OR_GOTO_FAIL(
			    assign_alloc(&f->value, obj->instream,
			        strdup(obj->instream->string)));
			return f;
		case TOKEN_OPEN_BRACE:
			f->type = YOCTON_FIELD_OBJECT;
			CHECK_OR_GOTO_FAIL(
			    assign_alloc(&f->child, obj->instream,
			        calloc(1, sizeof(struct yocton_object))));
			f->child->instream = obj->instream;
			f->child->done = 0;
			return f;
		default:
			input_error(obj->instream, "':' or '{' expected to "
			            "follow field name");
			goto fail;
	}
fail:
	free_field(f);
	return NULL;
}

struct yocton_field *yocton_next_field(struct yocton_object *obj)
{
	if (obj == NULL || obj->done || strlen(obj->instream->error_buf) > 0) {
		return NULL;
	}

	skip_forward(obj);
	free_field(obj->field);
	obj->field = NULL;

	switch (read_next_token(obj->instream)) {
		case TOKEN_STRING:
			return next_field(obj);
		case TOKEN_CLOSE_BRACE:
			if (obj == obj->instream->root) {
				input_error(obj->instream, "closing brace "
				            "'}' not expected at top level");
				return NULL;
			}
			obj->done = 1;
			return NULL;
		case TOKEN_EOF:
			// EOF is only valid at the top level.
			if (obj != obj->instream->root) {
				input_error(obj->instream, ERROR_EOF);
				return NULL;
			}
			obj->done = 1;
			return NULL;
		default:
			input_error(obj->instream, "expected start of "
			            "next field");
			return NULL;
	}
}

enum yocton_field_type yocton_field_type(struct yocton_field *f)
{
	return f->type;
}

const char *yocton_field_name(struct yocton_field *f)
{
	return f->name;
}

const char *yocton_field_value(struct yocton_field *f)
{
	if (f->type != YOCTON_FIELD_STRING) {
		input_error(f->parent->instream, "field '%s' has object, "
		            "not value type", f->name);
		return "";
	}
	return f->value;
}

struct yocton_object *yocton_field_inner(struct yocton_field *f)
{
	if (f->type != YOCTON_FIELD_OBJECT) {
		input_error(f->parent->instream, "field '%s' has value, "
		            "not object type", f->name);
		return NULL;
	}
	return f->child;
}

File: yocton.py
#!/usr/bin/env python
#
# Copyright (c) 2023, Simon Howard
#
# Permission to use, copy, modify, and/or distribute this software
# for any purpose with or without fee is hereby granted, provided
# that the above copyright notice and this permission notice appear
# in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
# WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
# AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
# CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
# LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

from __future__ import unicode_literals, print_function, generators
import collections
import io
import re

VALID_SYMBOL_RE = re.compile(r"^[a-zA-Z0-9_\-\+\.]+$")

TOKEN_STRING = 0
TOKEN_COLON = 1
TOKEN_OPEN_BRACE = 2
TOKEN_CLOSE_BRACE = 3
TOKEN_EOF = 4
TOKEN_NONE = 5

ASCII_X = ord('x')
ASCII_QUOTES = ord('"')
ASCII_SLASH = ord('/')
ASCII_BACKSLASH = ord('\\')
ASCII_NEWLINE = ord('\n')
ASCII_OPEN_BRACE = ord('{')
ASCII_CLOSE_BRACE = ord('}')
ASCII_COLON = ord(':')
ASCII_AMPERSAND = ord('&')
SIMPLE_ESCAPES = { ord('n'): ASCII_NEWLINE, ord('t'): ord('\t'),
                   ASCII_QUOTES: ASCII_QUOTES,
                   ASCII_BACKSLASH: ASCII_BACKSLASH }
REVERSE_ESCAPES = { "\n": "\\n", "\t": "\\t", '"': "\\\"", "\\": "\\\\" }
BOM = 0xfeff
UTF8_BOM = b'\xef\xbb\xbf'
ERROR_EOF = "unexpected EOF"

def is_symbol_char(c):
	return VALID_SYMBOL_RE.match("%c" % c) is not None

class Lexer(object):
	def __init__(self, instream):
		self.instream = instream
		self.buf = ''
		self.buf_index = 0
		self.lineno = 1
		self.root = None
		self.binary = False
		self.error = False

	def syntax_assert(self, error_if_false, error_message="syntax error",
	                  *args):
		if not error_if_false:
			self.error = True
			raise SyntaxError(error_message % args)

	def refill_buffer(self):
		"""Refill read buffer if it's empty; returns False for EOF."""
		if self.buf_index < len(self.buf):
			return True
		self.buf = self.instream.read(128)
		if len(self.buf) == 0:
			return False
		self.binary = not isinstance(self.buf, str)
		self.buf_index = 0
		return True

	def peek_next_char(self):
		"""Inspect the next input character without advancing.

		Returns Unicode or byte value representing character.
		"""
		if not self.refill_buffer():
			return False, None
		result = self.buf[self.buf_index]
		if not self.binary:
			result = ord(result)
		return True, result

	def read_next_char(self):
		"""Read next input character and advance.

		Returns integer representing next character.
		"""
		ok, c = self.peek_next_char()
		self.syntax_assert(ok, ERROR_EOF)
		self.buf_index += 1
		return c

	def read_escape_sequence(self):
		c = self.read_next_char()
		esc = SIMPLE_ESCAPES.get(c)
		if esc is not None:
			return esc
		self.syntax_assert(c == ASCII_X, "unknown string escape: \\%c", c)
		try:
			d1 = self.read_next_char()
			d2 = self.read_next_char()
			c = int("%c%c" % (d1, d2), 16)
		except ValueError:
			self.syntax_assert(False,
				"\\x sequence must be followed "
				"by two hexadecimal characters")
		self.syntax_assert(c != 0,
			"NUL byte not allowed in \\x escape sequence.")
		self.syntax_assert(c < 0x20,
			"\\x escape sequence can only be used for "
			"control characters (ASCII 0x01-0x1f range)")
		return c

	def to_string(self, s):
		if self.binary:
			return bytes(s)
		else:
			return ''.join(chr(c) for c in s)

	def _next_string_chunk(self):
		tt, c = self._skip_past_spaces()
		if tt == TOKEN_EOF:
			return TOKEN_STRING
		elif tt != TOKEN_NONE:
			return tt
		elif c != ASCII_AMPERSAND:
			return TOKEN_STRING
		self.read_next_char()

		tt, c = self._skip_past_spaces()
		self.syntax_assert(tt == TOKEN_NONE and c == ASCII_QUOTES,
			"quoted string should follow '&' operator")
		self.read_next_char()

		return TOKEN_NONE

	def read_string(self):
		result_chars = []
		while True:
			c = self.read_next_char()
			if c == ASCII_QUOTES:
				tt = self._next_string_chunk()
				if tt != TOKEN_NONE:
					break
				continue
			elif c == ASCII_BACKSLASH:
				c = self.read_escape_sequence()
			else:
				self.syntax_assert(c >= 0x20,
					'control character not allowed '
					'inside string (ASCII char 0x%02x)', c)
			result_chars.append(c)
		return self.to_string(result_chars)

	def read_symbol(self, first):
		self.syntax_assert(is_symbol_char(first),
			'unknown token: not valid symbol character')

		result_chars = [first]
		while True:
			ok, next_c = self.peek_next_char()
			if not ok or not is_symbol_char(next_c):
				# EOF is explicitly okay here
				break
			self.read_next_char()
			result_chars.append(next_c)

		return self.to_string(result_chars)

	def _skip_past_spaces(self):
		# Skip past spaces, including comments.
		while True:
			ok, c = self.peek_next_char()
			if not ok:
				return TOKEN_EOF, None
			if c == ASCII_SLASH:
				self.syntax_assert(
					self.read_next_char() == ASCII_SLASH and
					self.read_next_char() == ASCII_SLASH)
				while c != ASCII_NEWLINE:
					ok, c = self.peek_next_char()
					if not ok:
						return TOKEN_EOF, None
					self.read_next_char()
			elif self.binary and c == UTF8_BOM[0]:
				self.syntax_assert(
					self.read_next_char() == UTF8_BOM[0] and
					self.read_next_char() == UTF8_BOM[1] and
					self.read_next_char() == UTF8_BOM[2])
			elif not self.binary and c == BOM:
				self.read_next_char()
			elif not ("%c" % c).isspace():
				return TOKEN_NONE, c
			else:
				self.read_next_char()

	def do_read_next_token(self):
		"""The actual logic of read_next_token()."""
		tt, cc = self._skip_past_spaces()
		if tt != TOKEN_NONE:
			return tt, None
		c = self.read_next_char()
		if c == ASCII_COLON:
			return TOKEN_COLON, None
		elif c == ASCII_OPEN_BRACE:
			return TOKEN_OPEN_BRACE, None
		elif c == ASCII_CLOSE_BRACE:
			return TOKEN_CLOSE_BRACE, None
		elif c == ASCII_QUOTES:
			return TOKEN_STRING, self.read_string()
		else:
			self.syntax_assert(c != ASCII_AMPERSAND, "'&' operator"
				" can only be used to join quoted strings")
			return TOKEN_STRING, self.read_symbol(c)

	def read_next_token(self):
		if self.error:
			return TOKEN_EOF, None

		return self.do_read_next_token()

YoctonProperty = collections.namedtuple('YoctonProperty', ('name', 'value'))

class YoctonObject(object):
	def __init__(self, lexer, parent=None):
		self.lexer = lexer
		self.parent = parent
		self.child = None
		self.done = False
		if parent is None:
			lexer.root = self

	def __iter__(self):
		return self

	def __next__(self):
		p = self.next_prop()
		if p is None:
			raise StopIteration
		return p

	def next(self):
		return self.__next__()

	def skip_forward(self):
		while self.child is not None and not self.child.done:
			self.child.next_prop()
		self.child = None

	def parse_next_prop(self):
		tt, value = self.lexer.read_next_token()
		if tt == TOKEN_COLON:
			# This is the string:string case.
			tt, value = self.lexer.read_next_token()
			if tt != TOKEN_STRING:
				raise SyntaxError(
					"string expected to follow ':'")
			return value

		elif tt == TOKEN_OPEN_BRACE:
			self.child = YoctonObject(self.lexer, self)
			return self.child

		raise SyntaxError("':' or '{' expected to follow property name");

	def next_prop(self):
		if self.done:
			return None

		self.skip_forward()
		tt, value = self.lexer.read_next_token()

		if tt == TOKEN_STRING:
			return YoctonProperty(value, self.parse_next_prop())
		elif tt == TOKEN_CLOSE_BRACE:
			if self == self.lexer.root:
				raise SyntaxError("closing brace '}' not "
				                  "expected at top level")
			self.done = True
			return None
		elif tt == TOKEN_EOF:
			if self != self.lexer.root:
				raise EOFError(ERROR_EOF)
			self.done = True
			return None
		raise SyntaxError("expected start of next property")

class YoctonWriter(object):
	def __init__(self, outstream):
		self.outstream = outstream
		self.indent_level = 0

	def write_string(self, s):
		s = str(s)
		if VALID_SYMBOL_RE.match(s):
			self.outstream.write(s)
			return

		# TODO: Write bytes when in binary mode
		self.outstream.write('"')
		for c in s:
			esc = REVERSE_ESCAPES.get(c)
			if esc is not None:
				self.outstream.write(esc)
			elif ord(c) >= 0x20:
				self.outstream.write(c)
			else:
				self.outstream.write("\\x%02x" % ord(c))
		self.outstream.write('"')

	def write_indent(self):
		self.outstream.write("\t" * self.indent_level)

	def write_property(self, name, value):
		self.write_indent()
		self.write_string(name)
		self.outstream.write(": ")
		self.write_string(value)
		self.outstream.write("\n")
		# Always flush after top-level def
		if self.indent_level == 0:
			self.outstream.flush()

	def begin_subobject(self, name):
		self.write_indent()
		self.write_string(name)
		self.outstream.write(" {\n")
		self.indent_level += 1

	def end(self):
		if self.indent_level == 0:
			return
		self.indent_level -= 1
		self.write_indent()
		self.outstream.write("}\n")
		if self.indent_level == 0:
			self.outstream.flush()

def parse(fp):
	lexer = Lexer(fp)
	return YoctonObject(lexer)

def parses(s):
	if isinstance(s, str):
		buf = io.StringIO(s)
	else:
		buf = io.BytesIO(s)
	return parse(buf)

def dump(obj, fp):
	w = YoctonWriter(fp)

	def write_obj(o):
		if isinstance(o, dict):
			o = list(o.items())

		for name, value in o:
			if isinstance(value, (str, bytes, int)):
				w.write_property(name, value)
			else:
				w.begin_subobject(name)
				write_obj(value)
				w.end()

	write_obj(obj)

def dumps(obj):
	buf = io.StringIO()
	dump(obj, buf)
	buf.flush()
	return buf.getvalue()

def load(fp):
	def read_obj(o):
		result = []
		for name, value in o:
			if isinstance(value, YoctonObject):
				result.append((name, read_obj(value)))
			else:
				result.append((name, value))
		return result
	return read_obj(parse(fp))

def loads(s):
	if isinstance(s, str):
		buf = io.StringIO(s)
	else:
		buf = io.BytesIO(s)
	return load(buf)

